pkg/controller/controller_binding.go

```diff
@@ -1006,7 +1096,501 @@ func (c *controller) recordStartOfServiceBindingOperation(toUpdate *v1beta1.Serv
+		glog.V(4).Info(pcb.Message(s))
+		c.recorder.Event(binding, corev1.EventTypeWarning, errorPollingLastOperationReason, s)
+
+		if err := c.checkPollingServiceBindingForReconciliationRetryTimeout(binding); err != nil {
+			return nil
+		}
+
+		return c.continuePollingServiceBinding(binding)
+	}
+
+	glog.V(4).Info(pcb.Messagef("Poll returned %q : %q", response.State, response.Description))
+
+	switch response.State {
+	case osb.StateInProgress:
+		// if the description is non-nil, then update the instance condition with it
+		if response.Description != nil {
+			var (
+				message string
+				reason  string
+			)
+			if deleting {
+				reason = asyncUnbindingReason
+				message = asyncUnbindingMessage
+			} else {
+				reason = asyncBindingReason
+				message = asyncBindingMessage
+			}
+
+			message = fmt.Sprintf("%s (%s)", message, *response.Description)
+
+			setServiceBindingCondition(
+				binding,
+				v1beta1.ServiceBindingConditionReady,
+				v1beta1.ConditionFalse,
+				reason,
+				message,
+			)
+		}
+
+		if err := c.checkPollingServiceBindingForReconciliationRetryTimeout(binding); err != nil {
+			return nil
+		}
+
+		if _, err := c.updateServiceBindingStatus(binding); err != nil {
+			return err
+		}
+
+		glog.V(4).Info(pcb.Message("Last operation not completed (still in progress)"))
+
+		return c.continuePollingServiceBinding(binding)
+	case osb.StateSucceeded:
+		if deleting {
+			var (
+				reason  string
+				message string
+			)
+			switch {
+			case mitigatingOrphan:
+				reason = successOrphanMitigationReason
+				message = successOrphanMitigationMessage
+			default:
+				reason = successUnboundReason
+				message = "The binding was deleted successfully"
+			}
+
+			clearServiceBindingCurrentOperation(binding)
+			binding.Status.ExternalProperties = nil
+
+			setServiceBindingCondition(
+				binding,
+				v1beta1.ServiceBindingConditionReady,
+				v1beta1.ConditionFalse,
+				reason,
+				message,
+			)
+
+			if !mitigatingOrphan {
+				// Clear the finalizer
+				if finalizers := sets.NewString(binding.Finalizers...); finalizers.Has(v1beta1.FinalizerServiceCatalog) {
+					finalizers.Delete(v1beta1.FinalizerServiceCatalog)
+					binding.Finalizers = finalizers.List()
+				}
+			}
+
+			if _, err := c.updateServiceBindingStatus(binding); err != nil {
+				return err
+			}
+
+			c.recorder.Event(binding, corev1.EventTypeNormal, reason, message)
+			glog.V(4).Info(pcb.Message(message))
+
+			return c.finishPollingServiceBinding(binding)
+		}
+
+		// Update the in progress/external properties, as the changes have been
+		// persisted in the broker
+		binding.Status.ExternalProperties = binding.Status.InProgressProperties
+		binding.Status.InProgressProperties = nil
+
+		getBindingRequest := &osb.GetBindingRequest{
+			InstanceID: instance.Spec.ExternalID,
+			BindingID:  binding.Spec.ExternalID,
+		}
+
+		getBindingResponse, err := brokerClient.GetBinding(getBindingRequest)
+		if err != nil {
```

Comment:
Oof... I didn't notice the "MUST" there. I had intended to long-term move the GET outside of the polling since it is completely independent of the `last_operation` endpoint, but had hoped it could be a follow-up. Orphan mitigation would probably be the best short-term solution.

As an aside about polling in general, it seems problematic to explicitly say the platform MUST stop polling upon receiving a succeeded/failed state. Does that mean we should expect brokers to nuke their `last_operation` endpoint as soon as one GET that returns "succeeded" is returned? If so, then is our current instance polling is also a bit broken? If something goes wrong while trying to store the state of the `succeeded`/`failed` response, we try again from the top, attempting another poll.

Reply:
The alternative is that the broker would need to retain every `operation_key` indefinitely, though, right? Is there a happy medium where the brokers agree to retain the `operation_key` for a certain period of time or until the platform explicitly acknowledges that it no longer needs the `operation_key`?

Yes, I agree that our current polling is a bit broken. It is not clear (at least to me) from the OSB spec how the broker is supposed to respond when it gets a last operation. Either the broker will respond with an error code and the controller will continue polling until the reconciliation retry duration expires. Or the broker will respond with "succeeded" and the controller will proceed along merrily as though it is the first time that it received "succeeded". In the first case, there are quite a few edge cases that make that an undesirable state for the resource to be in.

Reply:
It seems to me as well like the spec is unclear. Rereading, the wording seems to imply to me that a platform should only expect one `"succeeded"`/`"failed"`, but as you said any error code other than 400 will cause the platform to keep polling, and the description of 400 doesn't capture this case. I'll bring it up to the OSB working group to see if we can get clarification.

For the time being, I made this situation trigger orphan mitigation.